Add `home` crate to cargo crates

LucioFranco opened the pull request:
### What does this PR try to resolve?

The `home` crate is the crate used by both `rustup` and `cargo` to define their home directories. The crate is quite small but plays an important role in gluing rust tooling together. This PR adds this crate the `crates` folder within cargo to allow the cargo team and the rust project to maintain this small crate.

### Additional information

I've also added both `rust-lang-owners` and `ehuss` as owners of the crate. Once, merged I will archive the old repo. 

Reference zulip conversation: https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Moving.20the.20home.20crate.20into.20rust-lang.20and.20under.20cargo/near/299975263

cc @weihanglo @ehuss

LucioFranco commented on crates/home/CHANGELOG.md:
```diff
@@ -0,0 +1,46 @@
+### Removed
+- Remove support for `multirust` folder used in old version of `rustup`.
+
+[Unreleased]: https://github.com/brson/home/compare/v0.5.4...HEAD
```
SO I think for now it makes sense to keep these linked to brson, but on the next release we can update the tag. Not sure what the process in cargo is. That said, I doubt we will need a release for a long time so something that  can be figured out then. The brson/home links should continue to work even if its archived iirc.

weihanglo replied:
There is no such a tagging policy for sub-crates in rust-lang/cargo. Tagging cargo itself is a separate process ran by Release team as well I believe (but currently broken 😅 https://github.com/rust-lang/cargo/issues/11345).

weihanglo commented on publish.py:
```diff
@@ -14,6 +14,7 @@
     'crates/cargo-platform',
     'crates/cargo-util',
     'crates/crates-io',
+    'crates/home',
```
@ehuss, I wonder whether we need to notify Release team for this change.

LucioFranco replied:
I didn't include this change in the second version since I don't think we should release home crate when we release cargo and this is the cargo release script I assume that runs when we do cargo specific releases?

weihanglo reviewed (approved):
Looks great. Thank you!

ehuss commented:
Thanks for the PR! I wasn't really expecting you to need to do any work here, so I appreciate it.

A few comments:
* I don't think this should be added to the publish script, or using a `path` dependency within cargo. Since this is a somewhat independent crate, I think we'll just manage it independently. I don't expect there needing to be too many coordinated updates between cargo and home, and if there are it shouldn't be too terrible to just publish a new version.
* I was wondering if it would be possible to import the tree with the full history? Perhaps something like subtree can assist with that process, or something like that?

LucioFranco commented:
> I don't think this should be added to the publish script, or using a path dependency within cargo. Since this is a somewhat independent crate, I think we'll just manage it independently. I don't expect there needing to be too many coordinated updates between cargo and home, and if there are it shouldn't be too terrible to just publish a new version.

Seems like a totally reasonable approach, will make this update.

> I was wondering if it would be possible to import the tree with the full history? Perhaps something like subtree can assist with that process, or something like that?

I was thinking about this but I didn't know if you wanted to make the cargo tree dirty from the home crates git history. My though process was that if we archive the current home repo we keep the history there and have a clean one going forward for cargo. That said, I am happy to make this change so I will explore doing that and will make that change as well.

LucioFranco commented:
@ehuss I think I've addressed all your concerns and this should be ready to go (pending CI)! Let me know if there is anything else you would like to see changed.

weihanglo reviewed (approved):
Looks good to me. Thanks!

ehuss commented on .github/workflows/main.yml:
```diff
@@ -99,6 +99,7 @@ jobs:
         CARGO_TARGET_DIR: target
     - run: cargo test -p cargo-platform
     - run: cargo test -p cargo-util
+    - run: cargo test -p home
```
I don't think this line is doing what the intended effect is. This is testing `home` from `crates-io`, not the one in this repository. The reason the other two work with `-p` is because they are path-dependencies. 

I think the fix here is to use something like `--manifest-path crates/home/Cargo.toml`

LucioFranco replied:
Good point, made that change in https://github.com/rust-lang/cargo/pull/11359/commits/18a06cc809cd0c7dc2a7297ad4a2b4b98b829d22

ehuss reviewed (commented):
Thanks for looking into how to bring in the history!

ehuss commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit 18a06cc809cd0c7dc2a7297ad4a2b4b98b829d22 has been approved by `ehuss`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=ehuss 18a06cc809cd0c7dc2a7297ad4a2b4b98b829d22 -->
<!-- homu: {"type":"Approved","sha":"18a06cc809cd0c7dc2a7297ad4a2b4b98b829d22","approver":"ehuss","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

weihanglo commented:
stuck?

@bors retry

ehuss commented:
Oh sorry, I should have commented somewhere. There was a GitHub outage this morning, and bors is now stuck. We have to wait for infra to fix it (or for bors to fix itself).

weihanglo commented:
@bors ping

bors commented:
:sleepy: I'm awake I'm awake

weihanglo commented:
@bors retry

weihanglo commented:
Still not fixed? I saw some PR in rust-lang/rust started to run.

@bors r- 
@bors r=ehuss

bors commented:
:pushpin: Commit 18a06cc809cd0c7dc2a7297ad4a2b4b98b829d22 has been approved by `ehuss`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=ehuss 18a06cc809cd0c7dc2a7297ad4a2b4b98b829d22 -->
<!-- homu: {"type":"Approved","sha":"18a06cc809cd0c7dc2a7297ad4a2b4b98b829d22","approver":"ehuss","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/3695398068/jobs/6258863264)
Approved by: ehuss
Pushing 06178d7a7c35cc2c058a4cf9391a21a4c5662eb0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ehuss","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/3695398068/jobs/6258863264"},"merge_sha":"06178d7a7c35cc2c058a4cf9391a21a4c5662eb0"} -->
